File: atmosphere.py
"""
Atmosphere model: humidity-corrected air density and speed of sound.

Humid air is LESS dense than dry air at the same pressure and temperature,
because water vapor (M = 18.02 g/mol) is lighter than the dry-air average
(28.9647 g/mol). Ignoring humidity therefore biases drop/drift high in hot,
humid conditions. We compute vapor pressure with the Buck (1996) equation and
mix partial pressures the CIPM way.
"""

import math
from dataclasses import dataclass

# Molar masses (kg/mol) and universal gas constant (J/mol/K)
M_DRY = 0.0289647
M_VAPOR = 0.0180160
R_UNIV = 8.314462618

# ICAO sea-level standard, used as the reference for stability corrections.
STD_TEMP_C = 15.0
STD_PRESSURE_PA = 101325.0
STD_RH = 0.0


def saturation_vapor_pressure_pa(temp_c: float) -> float:
    """Buck (1996) saturation vapor pressure over liquid water, in pascals."""
    return 100.0 * 6.1121 * math.exp(
        (18.678 - temp_c / 234.5) * (temp_c / (257.14 + temp_c))
    )


def air_density(temp_c: float, pressure_pa: float, humidity: float = 0.0) -> float:
    """
    Moist-air mass density (kg/m^3).

    temp_c       : ambient temperature, Celsius
    pressure_pa  : ABSOLUTE (station) pressure, pascals -- not sea-level-corrected
    humidity     : relative humidity as a fraction 0..1
    """
    p_v = humidity * saturation_vapor_pressure_pa(temp_c)
    p_d = pressure_pa - p_v
    t_k = temp_c + 273.15
    return (p_d * M_DRY + p_v * M_VAPOR) / (R_UNIV * t_k)


def speed_of_sound(temp_c: float, pressure_pa: float, humidity: float = 0.0) -> float:
    """
    Speed of sound in moist air (m/s).

    Humidity raises c slightly (lighter mixture); the effect is a few tenths of a
    percent but shifts the Mach number a bullet sees near transonic, so we keep it.
    """
    p_v = humidity * saturation_vapor_pressure_pa(temp_c)
    x_v = p_v / pressure_pa                      # mole fraction water vapor
    m_mix = (1.0 - x_v) * M_DRY + x_v * M_VAPOR
    # Ratio of specific heats: mole-fraction blend of dry air (1.400) and vapor (1.330)
    gamma = (1.0 - x_v) * 1.400 + x_v * 1.330
    t_k = temp_c + 273.15
    return math.sqrt(gamma * R_UNIV * t_k / m_mix)


def standard_density() -> float:
    return air_density(STD_TEMP_C, STD_PRESSURE_PA, STD_RH)


@dataclass
class Atmosphere:
    temp_c: float = 15.0
    pressure_pa: float = 101325.0
    humidity: float = 0.0          # fraction 0..1

    @property
    def density(self) -> float:
        return air_density(self.temp_c, self.pressure_pa, self.humidity)

    @property
    def mach1(self) -> float:
        return speed_of_sound(self.temp_c, self.pressure_pa, self.humidity)

File: atmosphere.rs
//! Air density and speed of sound: ISA, or measured station conditions with humidity.

use crate::math;

#[derive(Clone, Copy, Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Air {
    /// kg/m³
    pub density: f64,
    /// m/s
    pub speed_of_sound: f64,
}

/// Air as a function of height above the launch point (m).
pub trait Atmosphere {
    fn at(&self, height: f64) -> Air;
}

pub const R_DRY: f64 = 287.058; // J/(kg·K)
pub const R_VAPOR: f64 = 461.495; // J/(kg·K)
pub const GAMMA: f64 = 1.4;
/// ISA tropospheric lapse rate, K/m.
pub const LAPSE: f64 = 0.0065;
/// Standard gravity, m/s².
pub const G0: f64 = 9.80665;
pub const ISA_T0: f64 = 288.15; // K
pub const ISA_P0: f64 = 101_325.0; // Pa
/// ICAO sea-level density, kg/m³.
pub const ISA_RHO0: f64 = 1.225;
/// Army Standard Metro density (59 °F, 29.53 inHg, 78 % RH), kg/m³.
pub const ASM_RHO: f64 = 1.2030;

/// Saturation vapour pressure over water, Pa (Alduchov & Eskridge 1996).
pub fn saturation_vapor_pressure(temperature: f64) -> f64 {
    let tc = temperature - 273.15;
    610.94 * math::exp(17.625 * tc / (tc + 243.04))
}

/// Moist-air density and speed of sound from temperature (K), absolute
/// pressure (Pa) and relative humidity (0..1).
pub fn air(temperature: f64, pressure: f64, relative_humidity: f64) -> Air {
    let rh = relative_humidity.clamp(0.0, 1.0);
    let e = (rh * saturation_vapor_pressure(temperature)).min(pressure);
    let pd = pressure - e;
    let density = pd / (R_DRY * temperature) + e / (R_VAPOR * temperature);
    let speed_of_sound = math::sqrt(GAMMA * pressure / density);
    Air {
        density,
        speed_of_sound,
    }
}

/// ISA temperature (K) and pressure (Pa) at geometric altitude (m).
/// Troposphere model — good to 11 km, which is all a rifle needs.
pub fn isa_tp(altitude: f64) -> (f64, f64) {
    let t = ISA_T0 - LAPSE * altitude;
    let p = ISA_P0 * math::pow(t / ISA_T0, G0 / (R_DRY * LAPSE));
    (t, p)
}

/// Sea-level-corrected pressure (what a weather report gives) to absolute
/// station pressure at `altitude` m, using the ISA lapse.
pub fn sea_level_to_station(sea_level_pressure: f64, altitude: f64) -> f64 {
    let (t, _) = isa_tp(altitude);
    sea_level_pressure * math::pow(t / ISA_T0, G0 / (R_DRY * LAPSE))
}

/// ICAO standard atmosphere, dry. `station_altitude` is the launch point's
/// altitude ASL (m); heights passed to `at` are relative to it.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Isa {
    pub station_altitude: f64,
}

impl Atmosphere for Isa {
    fn at(&self, height: f64) -> Air {
        let (t, p) = isa_tp(self.station_altitude + height);
        air(t, p, 0.0)
    }
}

/// Measured conditions at the launch point (Kestrel-style). Above/below the
/// launch point, temperature and pressure follow the ISA lapse from the
/// measured values; humidity is held constant.
#[derive(Clone, Copy, Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Station {
    /// K
    pub temperature: f64,
    /// Absolute (station) pressure, Pa. Not sea-level corrected.
    pub pressure: f64,
    /// 0..1
    pub humidity: f64,
}

impl Atmosphere for Station {
    fn at(&self, height: f64) -> Air {
        let t = self.temperature - LAPSE * height;
        let p = self.pressure * math::pow(t / self.temperature, G0 / (R_DRY * LAPSE));
        air(t, p, self.humidity)
    }
}
